Mathlib/Combinatorics/Schnirelmann.lean

```diff
@@ -0,0 +1,247 @@
+the positive naturals. -/
+noncomputable def schnirelmannDensity (A : Set ℕ) [DecidablePred (· ∈ A)] : ℝ :=
+  ⨅ n : {n : ℕ // 0 < n}, ((Ioc (0 : ℕ) n).filter (· ∈ A)).card / n
+
+section
+
+variable {A : Set ℕ} [DecidablePred (· ∈ A)]
+
+lemma schnirelmannDensity_nonneg : 0 ≤ schnirelmannDensity A :=
+  Real.iInf_nonneg (fun _ => by positivity)
+
+lemma schnirelmannDensity_le_div {n : ℕ} (hn : n ≠ 0) :
+    schnirelmannDensity A ≤ ((Ioc 0 n).filter (· ∈ A)).card / n :=
+  ciInf_le ⟨0, fun _ ⟨_, hx⟩ => hx ▸ by positivity⟩ (⟨n, hn.bot_lt⟩ : {n : ℕ // 0 < n})
+
+/--
+For any natural `n`, the Schnirelmann density multiplied by `n` is bounded by `|A ∩ {1, ..., n}|`.
+Note this property fails for the natural density.
+-/
+lemma schnirelmannDensity_mul_le_card_filter {n : ℕ} :
+    schnirelmannDensity A * n ≤ ((Ioc 0 n).filter (· ∈ A)).card := by
+  rcases eq_or_ne n 0 with rfl | hn
+  · simp
+  exact (le_div_iff (by positivity)).1 (schnirelmannDensity_le_div hn)
+
+/--
+To show the Schnirelmann density is upper bounded by `x`, it suffices to show
+`|A ∩ {1, ..., n}| / n ≤ x`, for any chosen positive value of `n`.
+
+We provide `n` explicitly here to make this lemma more easily usable in `apply` or `refine`.
+This lemma is analogous to `ciInf_le_of_le`.
+-/
+lemma schnirelmannDensity_le_of_le {x : ℝ} (n : ℕ) (hn : n ≠ 0)
+    (hx : ((Ioc 0 n).filter (· ∈ A)).card / n ≤ x) : schnirelmannDensity A ≤ x :=
+  (schnirelmannDensity_le_div hn).trans hx
+
+lemma schnirelmannDensity_le_one : schnirelmannDensity A ≤ 1 :=
+  schnirelmannDensity_le_of_le 1 one_ne_zero <|
+    by rw [Nat.cast_one, div_one, Nat.cast_le_one]; exact card_filter_le _ _
+
+/--
+If `k` is omitted from the set, its Schnirelmann density is upper bounded by `1 - 1 / k`.
+-/
+lemma schnirelmannDensity_le_of_not_mem {k : ℕ} (hk : k ∉ A) :
+    schnirelmannDensity A ≤ 1 - 1 / k := by
+  rcases k.eq_zero_or_pos with rfl | hk'
+  · simpa using schnirelmannDensity_le_one
+  apply schnirelmannDensity_le_of_le k hk'.ne'
+  rw [one_sub_div (Nat.cast_pos.2 hk').ne']
+  apply div_le_div_of_le (Nat.cast_nonneg _)
+  rw [←Nat.cast_pred hk', Nat.cast_le]
+  suffices : (Ioc 0 k).filter (· ∈ A) ⊆ Ioo 0 k; exact (card_le_of_subset this).trans_eq (by simp)
+  rw [←Ioo_insert_right hk', filter_insert, if_neg hk]
+  exact filter_subset _ _
+
+/-- The Schnirelmann density of a set not containing `1` is `0`. -/
+lemma schnirelmannDensity_eq_zero_of_one_not_mem (h : 1 ∉ A) : schnirelmannDensity A = 0 :=
+  ((schnirelmannDensity_le_of_not_mem h).trans (by simp)).antisymm schnirelmannDensity_nonneg
+
+/-- The Schnirelmann density is increasing with the set. -/
+lemma schnirelmannDensity_le_of_subset {B : Set ℕ} [DecidablePred (· ∈ B)] (h : A ⊆ B) :
+    schnirelmannDensity A ≤ schnirelmannDensity B :=
+  ciInf_mono ⟨0, fun _ ⟨_, hx⟩ => hx ▸ by positivity⟩ fun _ => div_le_div_of_le (by positivity) <|
+    Nat.cast_le.2 <| card_le_of_subset <| monotone_filter_right _ h
```

Comment:
```suggestion
  ciInf_mono ⟨0, fun _ ⟨_, hx⟩ => hx ▸ by positivity⟩ fun _ => by
    gcongr; exact monotone_filter_right _ h
```

Reply:
This is a particular example I'm worried about when it comes to non-terminal `gcongr` - where adding new things to `gcongr`could produce errors here...